add alpha as method and as a cycle condition

RaulPROP opened the pull request:
I have added a new cycle conditioner called `cooldownAlpha` that checks whether the current alpha of the system is greater or equal to the one given, and stop the engine if it isn't.

Also, I have added a method `alpha` no access that value.

The idea behind this is to stop the engine when the data is in a state when it is unlikely to move, rather than a fixed amount of time / frames.

By default that value is negative infinite so it doesn't affect the other conditions.

(I have also added a [PR to the 3d-force-graph library](https://github.com/vasturiano/3d-force-graph/pull/309))

RaulPROP committed:
add alpha as method and as a cycle condition

vasturiano commented:
@RaulPROP I think this functionality is identical to the `alphaMin` property of the `d3-force-3d` engine:
https://github.com/vasturiano/d3-force-3d#simulation_alphaMin

So perhaps it would be simpler to expose that instead, similar to how `d3AlphaDecay` is setup:
https://github.com/vasturiano/three-forcegraph/blob/master/src/forcegraph-kapsule.js#L158

RaulPROP commented:
Oh, cool, didn't see that. Then yes, maybe that's the proper way to expose this functionality.

Should I add a commit modifying it?

vasturiano commented:
Actually, I just realized that the built-in d3-force `alphaMin` won't really work because that's only when the simulation is running automatically. In our case we're running it manually using [simulation.tick()](https://github.com/vasturiano/d3-force-3d#simulation_tick) so the `alphaMin` will not function to stop the simulation.

Thus, I think it's worth keeping the name of `d3AlphaMin` for this new property, but the implementation of it should be close to what you have now. The one thing that's missing in your PR is checking whether it's using `d3` as the `forceEngine`, otherwise it shouldn't stop the engine (because `ngraph` doesn't have an `alpha` param to compare against). Using the name of `d3AlphaMin` also makes this more clear.

RaulPROP commented:
Doesn't `ngraph` has some sort of "energy" (how much has the nodes move from the previous iteration)? We could use that instead of the alpha. (I haven't use `ngraph` so I don't know)

And yes, `d3AlphaMin` sound good, and it would work both on "warmup" and "cooldown" iterations, so there is no misconception with the "cooldown" suffix.

vasturiano commented:
`ngraph`'s settings are not compatible with `d3-force` so I wouldn't attempt to mix them.

Sounds good to also add it to the warmup phase, but you'll need to extend the iteration code here:
https://github.com/vasturiano/three-forcegraph/blob/master/src/forcegraph-kapsule.js#L1029

Also, would be great if we could apply the same functionality to the 2D variant of this, to maintain feature parity. 😄 
https://github.com/vasturiano/force-graph

It follows a similar overall code structure.

RaulPROP committed:
modified alpha naming and checking enigne

RaulPROP commented:
I have made those changes, which are:
- Rename `cooldownAlpha` to `d3AlphaMin`.
- Rename `alpha` to `d3Alpha`.
- Use `d3AlphaMin` in the warmup iterations.

The only thing that I'm not so sure is that i added `d3Alpha` to the methods array, because that value changes with each iteration. I'm not sure if that is correct.

Also added [PR to force-graph](https://github.com/vasturiano/force-graph/pull/117) and [modified the 3d-force-graph one](https://github.com/vasturiano/3d-force-graph/pull/309/commits/499c06f2a0436b8c8f2fcab4cd5ee6889cea60ca).

Off-topic: It would be nice if you could add a code formatter config file, because mine was messing with the code, changing every line of code.

vasturiano commented on README.md:
```diff
@@ -84,6 +84,8 @@ myGraph.tickFrame();
 | <b>emitParticle</b>(<i>link</i>) | An alternative mechanism for generating particles, this method emits a non-cyclical single particle within a specific link. The emitted particle shares the styling (speed, width, color) of the regular particle props. A valid `link` object that is included in `graphData` should be passed as a single parameter. ||
 | <b>forceEngine</b>([<i>str</i>]) | Getter/setter for which force-simulation engine to use ([*d3*](https://github.com/vasturiano/d3-force-3d) or [*ngraph*](https://github.com/anvaka/ngraph.forcelayout)). | `d3` |
 | <b>d3ReheatSimulation</b>() | Reheats the force simulation engine, by setting the `alpha` value to `1`. Only applicable if using the d3 simulation engine. | |
+| <b>d3Alpha</b>([<i>num</i>]) | Getter for the [simulation alpha](https://github.com/vasturiano/d3-force-3d#simulation_alpha) parameter, only applicable if using the d3 simulation engine. | `1` |
```
I don't think we need to expose `d3Alpha` externally. That should remain an internal parameter of the engine. One should be able to calibrate the engine behavior strictly from the alpha decay and min properties.

RaulPROP replied:
They idea behind it was to have a getter so the user can have more information about the state of the graph.

vasturiano replied:
From the doc line it looks like a setter (you can specify a `number`) which is disadvisable. I'm still leaning towards not exposing this prop since it's really a system internal.

RaulPROP replied:
Actually the only reason i added it was so in [the 3d-force-graph PR](https://github.com/vasturiano/3d-force-graph/pull/309) the user could see that value, like kind of a very poor test.

I don't mind removing it, but also it doesn't harm the user having that extra information.

vasturiano replied:
Not to snowball the API, prefer to leave it out until there's a compelling use case.

vasturiano commented on src/forcegraph-kapsule.js:
```diff
@@ -207,7 +208,11 @@ export default Kapsule({
+        if (
+          ++state.cntTicks > state.cooldownTicks ||
+          (new Date()) - state.startTickTime > state.cooldownTime ||
+          state.d3ForceLayout.alpha() < state.d3AlphaMin
```
As mentioned previously we need to check whether it's using the D3 layout (`isD3Sim`) before checking this condition.

vasturiano commented on src/forcegraph-kapsule.js:
```diff
@@ -531,7 +536,10 @@ export default Kapsule({
       return this;
-    }
+    },
+    d3Alpha: function (state) {
```
See above. No need to expose this method.

vasturiano commented on src/forcegraph-kapsule.js:
```diff
@@ -1026,7 +1034,14 @@ export default Kapsule({
+      for (
+        let i = 0;
+        i < state.warmupTicks &&
+        !(isD3Sim && (state.d3ForceLayout.alpha() < state.d3AlphaMin));
```
For not affecting backwards performance, I'd suggest to check if `d3AlphaMin > 0` before comparing it with the d3 alpha.

RaulPROP replied:
Do you mean before the hole block
```
!(isD3Sim && (state.d3ForceLayout.alpha() < state.d3AlphaMin))
```

or just the second statement?

```
(state.d3ForceLayout.alpha() < state.d3AlphaMin)
```


And I guess that it should also be checked in the layoutTick, right?

vasturiano replied:
I'd do 
```js
!(isD3Sim && state.d3AlphaMin > 0 && state.d3ForceLayout.alpha() < state.d3AlphaMin)
```

And yes, it should be the same check on the layoutTick.

RaulPROP committed:
remove d3Alpha method

RaulPROP commented:
Added the following changes:
- Removed d3Alpha method
- Modified cooldown cycle condition to check d3Sim
- Added state.d3AlphaMin > 0 condition

If everything is fine I will update `force-graph` and `3d-force-graph` accordingly.

vasturiano commented on src/forcegraph-kapsule.js:
```diff
@@ -155,6 +155,7 @@ export default Kapsule({
     linkDirectionalParticleColor: {},
     linkDirectionalParticleResolution: { default: 4 }, // how many slice segments in the particle sphere's circumference
     forceEngine: { default: 'd3' }, // d3 or ngraph
+    d3AlphaMin: { default: Number.NEGATIVE_INFINITY, triggerUpdate: false, onChange(alphaMin, state) { state.d3ForceLayout.alphaMin(alphaMin) }},
```
Let's set the default to `0`, since negative values of alpha don't really make sense, and `0` refers to the current backwards compatible behavior.
Also update the default in the docs.

vasturiano commented on src/forcegraph-kapsule.js:
```diff
@@ -531,7 +536,7 @@ export default Kapsule({
 
       return this;
-    }
+    },
```
Trailing comma

RaulPROP committed:
modified alphaMin initial value

RaulPROP commented:
Right, so:
- Removed trailing comma
- Set `d3AlphaMin` initial value to 0

vasturiano commented:
Thanks again for the PR @RaulPROP. I've done just some minor fixes and extended the type definitions. Final merge is here:
https://github.com/vasturiano/three-forcegraph/commit/9469a3f53a5451288c5cd3784aa13511e777abac

In order to fully expose this prop to all consuming modules, there's a few repos that would need updating and docs extended:
https://github.com/vasturiano/force-graph
https://github.com/vasturiano/3d-force-graph
https://github.com/vasturiano/aframe-forcegraph-component
https://github.com/vasturiano/3d-force-graph-vr
https://github.com/vasturiano/3d-force-graph-ar
https://github.com/vasturiano/react-force-graph

This is quite a bit of work but hopefully you can do at least `force-graph` and `3d-force-graph` and I'll try to take care of the rest. 👍

RaulPROP commented:
I have updated the [`3d-force-graph`](https://github.com/vasturiano/3d-force-graph/pull/309) and [`force-graph`](https://github.com/vasturiano/force-graph/pull/117) PRs.

I will try to do the same for the rest tomorrow :)
